precompile/assetsbridge/bridge_out.go

```diff
@@ -0,0 +1,620 @@
+		// TODO(Jeremy): Should we handle the revert properly
+		// to return an error message? For now just returning
+		// a generic error message
+		return fmt.Errorf("ERC20 balanceOf call reverted")
+	}
+
+	balance, err := extractBigIntFromEVMResult(resp.Ret)
+	if err != nil {
+		return fmt.Errorf("unable to unpack EVM result: %v", err)
+	}
+
+	// if the amount <= to the balance, then OK
+	// else we do not have enough funds and we need
+	// to return an error
+	if amount.Cmp(balance) > 0 {
+		return fmt.Errorf("not enough funds to bridgeOut")
+	}
+
+	return nil
+}
+
+func (m *BridgeOutMethod) isValidToken(
+	sdkCtx sdk.Context,
+	token common.Address,
+	chain TargetChain,
+) bool {
+	if chain == TargetChainEthereum { // both BTC and supported ERC20 are valid
+		btcToken := common.HexToAddress(evmtypes.BTCTokenPrecompileAddress)
+		if bytes.Equal(btcToken.Bytes(), token.Bytes()) {
+			return true
+		}
+
+		if _, ok := m.bridgeKeeper.GetERC20TokenMapping(sdkCtx, token.Bytes()); ok {
+			return true
+		}
+	} else if chain == TargetChainBitcoin { // only BTC is valid
+		btcToken := common.HexToAddress(evmtypes.BTCTokenPrecompileAddress)
+		if bytes.Equal(btcToken.Bytes(), token.Bytes()) {
+			return true
+		}
+	}
+
+	return false
+}
+
+// extractInputs extract the inputs from the precompile.MethodInputs and apply
+// basic validation on the inputs
+func (m *BridgeOutMethod) extractInputs(inputs precompile.MethodInputs) (*bridgeOutInputs, error) {
+	if err := precompile.ValidateMethodInputsCount(inputs, 4); err != nil {
+		return nil, err
+	}
+
+	token, ok := inputs[0].(common.Address)
+	if !ok {
+		return nil, fmt.Errorf("invalid token address: %v", inputs[0])
+	}
+
+	amount, ok := inputs[1].(*big.Int)
+	if !ok {
+		return nil, fmt.Errorf("invalid amount: %v", inputs[1])
+	}
+
+	if amount == nil {
+		return nil, errors.New("amount is required")
+	}
+
+	if amount.Sign() <= 0 {
+		return nil, errors.New("amount must be positive")
+	}
+
+	chainRaw, ok := inputs[2].(uint8)
+	if !ok {
+		return nil, fmt.Errorf("invalid chain: %v", inputs[2])
+	}
+
+	chain, ok := TargetChain(chainRaw).Validate()
+	if !ok {
+		return nil, fmt.Errorf("unsupported chain: %v", inputs[2])
+	}
+
+	recipient, ok := inputs[3].([]byte)
+	if !ok {
+		return nil, fmt.Errorf("invalid recipient address: %v", inputs[3])
+	}
+
+	if err := m.validateRecipientForChain(chain, recipient); err != nil {
+		return nil, err
+	}
+
+	return &bridgeOutInputs{
+		Token:     token,
+		Amount:    amount,
+		Chain:     chain,
+		Recipient: recipient,
+	}, nil
+}
+
+func (m *BridgeOutMethod) validateRecipientForChain(chain TargetChain, recipient []byte) error {
+	if len(recipient) == 0 {
+		return fmt.Errorf("recipient can't be empty")
+	}
+
+	switch chain {
+	case TargetChainEthereum:
+		// here we just check the length, the zero address have been tested before already
+		if len(recipient) != 20 {
+			return fmt.Errorf("invalid recipient address format for Ethereum chain: %v", recipient)
```

Comment:
`recipient` is `[]byte`. We should probably encode it to hex before printing:
```
hex.EncodeToString(recipient)
```
We should do the same when printing for Bitcoin.

Reply:
35231d1e, not I haven't done anything for btc yet, as I'm not sure hex makes sense  in that case? Does it?

Reply:
Please see https://github.com/mezo-org/mezod/pull/509#discussion_r2222094217.

Reply:
done in: 6dc7c4ba5d7999e2ccde4d2358889881cae99aab